app/cluster/dynamic.go

```diff
@@ -62,33 +74,60 @@ func (d *Dynamic) init() {
 	d.serverStopTimeStat = stats.NewTaggedStat("cluster.server_stop_time", stats.TimerType, tag)
 	d.serverStartCountStat = stats.NewTaggedStat("cluster.server_start_count", stats.CountType, tag)
 	d.serverStopCountStat = stats.NewTaggedStat("cluster.server_stop_count", stats.CountType, tag)
+
+	if d.BackendConfig == nil {
+		d.BackendConfig = backendconfig.DefaultBackendConfig
+	}
 }
 
 func (d *Dynamic) Run(ctx context.Context) error {
-	d.once.Do(func() {
-		d.init()
-	})
-	defer d.Provider.Close()
-	modeChan, err := d.Provider.ServerMode()
-	if err != nil {
-		return err
-	}
+	d.once.Do(d.init)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	serverModeChan := d.Provider.ServerMode(ctx)
+	workspaceIDsChan := d.Provider.WorkspaceIDs(ctx)
+
 	for {
 		select {
 		case <-ctx.Done():
 			if d.currentMode == servermode.NormalMode {
 				d.stop()
```

Comment:
[question] First of all I'm sorry if this was already explained :sweat_smile: 
Can you explain why on context cancelation if the current mode is `NORMAL` we "stop" and if not (meaning it is `DEGRADED`, right?) then we don't stop but we still return? Does it mean that the processor and router will still be running?

Reply:
If the mode is `DEGRADED` by default it would mean that both processor and router as stopped